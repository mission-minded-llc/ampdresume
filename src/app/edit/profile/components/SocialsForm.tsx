import { Box, Button, Dialog, DialogContent, List, TextField, Typography } from "@mui/material";
import { FieldDescription, FieldTitle, InputSection, SectionTitle } from "./sections";
import {
  generateSocialUrl,
  getSocialIcon,
  getSocialMediaPlatformByHostname,
  getSocialMediaPlatformByPlatformName,
} from "@/util/social";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { CustomDialogTitle } from "@/components/CustomDialogTitle";
import { DeleteWithConfirmation } from "../../components/DeleteWithConfirmation";
import { Icon } from "@iconify/react";
import { LoadingOverlay } from "@/components/LoadingOverlay";
import { MuiLink } from "@/components/MuiLink";
import { SOCIAL_MEDIA_PLATFORMS } from "@/constants";
import { Social } from "@ampdresume/theme";
import { Tooltip } from "@/components/Tooltip";
import { addSocial } from "@/graphql/addSocial";
import { deleteSocial } from "@/graphql/deleteSocial";
import { getSocials } from "@/graphql/getSocials";
import { updateSocial } from "@/graphql/updateSocial";
import { useSession } from "next-auth/react";
import { useState } from "react";

export const SocialsForm = () => {
  const { status, data: session } = useSession();
  const queryClient = useQueryClient();

  const [isOpen, setIsOpen] = useState(false);

  const [editSocial, setEditSocial] = useState<Social | null>(null);
  const [newSocialUrl, setNewSocialUrl] = useState("");
  const [errorMessage, setErrorMessage] = useState("");

  const isAuthenticatedUser = status === "authenticated" && !!session?.user.id;

  const {
    isPending,
    error,
    data: socials,
  } = useQuery({
    enabled: isAuthenticatedUser,
    queryKey: ["socials"],
    queryFn: async () => await getSocials(session?.user.id),
  });

  const mutationAddSocial = useMutation({
    mutationFn: async ({ platform, ref }: { platform: string; ref: string }) => {
      if (!session?.user.id) return null;

      await addSocial({
        userId: session?.user.id,
        platform,
        ref,
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["socials"] });

      setNewSocialUrl("");
      setErrorMessage("");
    },
  });

  const mutationUpdateSocial = useMutation({
    mutationFn: async ({ id, ref }: { id: string; ref: string }) => {
      if (!session?.user.id) return null;

      await updateSocial({ userId: session?.user.id, id, ref });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["socials"] });
    },
  });

  const mutationDeleteSocial = useMutation({
    mutationFn: async ({ id }: { id: string }) => {
      if (!session?.user.id) return null;

      await deleteSocial({ userId: session?.user.id, id });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["socials"] });
    },
  });

  const handleSocialUrlChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    let url = event.target.value;

    // Prefix URL with https:// if it does not have a protocol.
    if (!url.includes("http://") && !url.includes("https://")) {
      url = `https://${url}`;
    }

    setNewSocialUrl(url);
    setErrorMessage("");
  };

  const addSocialEntry = () => {
    try {
      const url = new URL(newSocialUrl);

      const hostname = url.hostname.replace("www.", "");
      const platformDetails = getSocialMediaPlatformByHostname(hostname);

      let platform = "website";
      let ref = newSocialUrl;

      if (platformDetails.regex) {
        // Extract the username via platform.regex against the newSocialUrl.
        // If the regex does not match, throw an error.
        const usernameMatch = newSocialUrl.match(platformDetails.regex);
        if (!usernameMatch) {
          throw new Error("Username not found in URL");
        }

        ref = usernameMatch[1];
        platform = platformDetails.name.toLowerCase();
      }

      mutationAddSocial.mutate({ platform, ref });
    } catch {
      setErrorMessage("Error adding social, please check the URL.");
    }
  };

  const updateSocialEntry = () => {
    if (editSocial) {
      mutationUpdateSocial.mutate({ id: editSocial.id, ref: editSocial.ref });
    }

    setIsOpen(false);
  };

  const deleteSocialEntry = () => {
    if (editSocial) {
      mutationDeleteSocial.mutate({ id: editSocial.id });
      setIsOpen(false);
    }
  };

  if (status === "loading") return <LoadingOverlay message="Loading session..." />;
  if (status === "unauthenticated") return null;
  if (isPending) return <LoadingOverlay message="Loading resume data..." />;
  if (error) return <Box>Error loading socials: {error.message}</Box>;

  return (
    <>
      <SectionTitle>Social Media Links</SectionTitle>
      <InputSection>
        <FieldTitle component="span">
          Add Link{" "}
          <Tooltip
            message={
              <List>
                <Typography variant="body2" sx={{ mb: 2 }}>
                  Supported Social URLs:
                </Typography>
                {Object.values(SOCIAL_MEDIA_PLATFORMS).map((platform) => (
                  <Typography key={platform.name} variant="body2">
                    {platform.name}
                  </Typography>
                ))}
              </List>
            }
          />
        </FieldTitle>
        <FieldDescription>
          Paste a link below to add to your socials. Link icons will appear at the top of your
          resume.
        </FieldDescription>
        <TextField
          label="Social URL"
          name="newSocialUrl"
          value={newSocialUrl}
          onChange={handleSocialUrlChange}
          fullWidth
          sx={{ marginTop: "auto" }}
          error={!!errorMessage}
          helperText={errorMessage}
        />
        <Button variant="contained" color="primary" sx={{ marginTop: 2 }} onClick={addSocialEntry}>
          Add Social
        </Button>
      </InputSection>

      <Box>
        {socials && socials?.length > 0 ? (
          <>
            <FieldTitle>Current Socials</FieldTitle>
            <Box sx={{ display: "flex", flexDirection: "row", flexWrap: "wrap", gap: 3 }}>
              {socials.map((social) => (
                <Box
                  key={social.id}
                  onClick={() => {
                    setEditSocial(social);
                    setIsOpen(true);
                  }}
                  sx={{ cursor: "pointer" }}
                >
                  <Icon icon={getSocialIcon(social)} width="48" height="48" />
                </Box>
              ))}
            </Box>
          </>
        ) : null}
      </Box>

      <Dialog open={isOpen} onClose={() => setIsOpen(false)}>
        <CustomDialogTitle closeHandler={() => setIsOpen(false)}>Edit Social</CustomDialogTitle>
        <DialogContent>
          {editSocial ? (
            <Box sx={{ p: 2, display: "flex", flexDirection: "column", gap: 2 }}>
              <Typography variant="body2" sx={{ mb: 2 }}>
                View:{" "}
                <MuiLink target="_blank" href={generateSocialUrl(editSocial)}>
                  {generateSocialUrl(editSocial).split("://")[1]}
                </MuiLink>
              </Typography>
              <TextField
                label={`Edit ${getSocialMediaPlatformByPlatformName(editSocial.platform).name}`}
                value={editSocial.ref}
                onChange={(e) => {
                  const socialHandleRef = e.target.value.replaceAll("@", "");
                  setEditSocial({ ...editSocial, ref: socialHandleRef });
                }}
              />
              <Box
                sx={{
                  display: "flex",
                  flexDirection: "row",
                  justifyContent: "space-between",
                  gap: 2,
                }}
              >
                <DeleteWithConfirmation onConfirmDelete={deleteSocialEntry} />
                <Button variant="contained" color="primary" onClick={updateSocialEntry}>
                  Save
                </Button>
              </Box>
            </Box>
          ) : (
            "No social selected."
          )}
        </DialogContent>
      </Dialog>
    </>
  );
};
