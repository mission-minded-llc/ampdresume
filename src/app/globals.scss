@use "@/app/vars.scss" as *;

html,
body {
  max-width: 100vw;
  width: 100%;
  height: 100%;
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
}

a {
  color: inherit;
}

ul {
  padding-inline-start: 0;
}

h3,
h4 {
  margin-bottom: 10px;
}

// Small tablet and mobile.
@media screen and (max-width: $breakpoint_mobile) {
  h2 {
    margin: 20px 0;
    padding: 20px 0;
  }
}
