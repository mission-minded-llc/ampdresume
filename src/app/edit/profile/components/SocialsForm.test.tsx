import "@testing-library/jest-dom";
import { useSession } from "next-auth/react";
import React from "react";
import { render } from "@testing-library/react";
import { useQuery, useQueryClient } from "@tanstack/react-query";
import { SocialsForm } from "./SocialsForm";
import { expect } from "@jest/globals";

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("@tanstack/react-query", () => ({
  useQuery: jest.fn(),
  useMutation: jest.fn(),
  useQueryClient: jest.fn(),
}));

jest.mock("@/graphql/addSocial", () => ({
  addSocial: jest.fn(),
}));

jest.mock("@/graphql/updateSocial", () => ({
  updateSocial: jest.fn(),
}));

jest.mock("@/graphql/deleteSocial", () => ({
  deleteSocial: jest.fn(),
}));

jest.mock("@/components/CustomDialogTitle", () => ({
  CustomDialogTitle: ({
    children,
    closeHandler,
  }: {
    children: React.ReactNode;
    closeHandler: () => void;
  }) => (
    <div>
      <button onClick={closeHandler}>Close</button>
      {children}
    </div>
  ),
}));

jest.mock("@/components/Tooltip", () => ({
  Tooltip: ({ message }: { message: React.ReactNode }) => <div>{message}</div>,
}));

jest.mock("@/components/LoadingOverlay", () => ({
  LoadingOverlay: ({ message }: { message: string }) => <div>{message}</div>,
}));

jest.mock("@/components/MuiLink", () => ({
  MuiLink: ({ children, ...props }: { children: React.ReactNode }) => <a {...props}>{children}</a>,
}));

describe("SocialsForm", () => {
  const mockSession = {
    user: { id: "user-id" },
  };

  const mockQueryClient = {
    invalidateQueries: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (useSession as jest.Mock).mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });
    (useQueryClient as jest.Mock).mockReturnValue(mockQueryClient);
  });

  it("renders correctly", () => {
    (useQuery as jest.Mock).mockReturnValue({ isPending: false, data: [] });

    const { container, getByLabelText, getByText } = render(<SocialsForm />);

    expect(getByLabelText("Social URL")).toBeInTheDocument();
    expect(getByText("Add Social")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("displays loading overlay when socials are being fetched", () => {
    (useQuery as jest.Mock).mockReturnValue({ isPending: true });

    const { getByText } = render(<SocialsForm />);
    expect(getByText("Loading resume data...")).toBeInTheDocument();
  });

  it("displays error message when socials fetching fails", () => {
    (useQuery as jest.Mock).mockReturnValue({
      isPending: false,
      error: { message: "Error" },
    });

    const { getByText } = render(<SocialsForm />);
    expect(getByText("Error loading socials: Error")).toBeInTheDocument();
  });
});
