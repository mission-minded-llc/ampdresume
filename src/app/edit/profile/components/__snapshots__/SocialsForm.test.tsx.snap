// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialsForm renders correctly 1`] = `
<div>
  <h2
    class="MuiTypography-root MuiTypography-h5 css-xt72g8-MuiTypography-root"
  >
    Social Media Links
  </h2>
  <div
    class="MuiBox-root css-18mctc4"
  >
    <span
      class="MuiTypography-root MuiTypography-h6 css-115ey5t-MuiTypography-root"
    >
      Add Link
       
      <div>
        <ul
          class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-956xpl-MuiTypography-root"
          >
            Supported Social URLs:
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            X
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            LinkedIn
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            GitHub
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Facebook
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Instagram
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            YouTube
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            TikTok
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Twitch
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Snapchat
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Reddit
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            WhatsApp
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Telegram
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Signal
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Medium
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
          >
            Website
          </p>
        </ul>
      </div>
    </span>
    <p
      class="MuiTypography-root MuiTypography-body2 css-1a1whku-MuiTypography-root"
    >
      Paste a link below to add to your socials. Link icons will appear at the top of your resume.
    </p>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1vc4ixa-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Social URL
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          name="newSocialUrl"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-81qg8w"
          >
            <span>
              Social URL
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-5pf1h4-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Add Social
    </button>
  </div>
  <div
    class="MuiBox-root css-0"
  />
</div>
`;
